Pattern Lock Added

sd1998 opened the pull request:
Fix to Issue #2077 

_____

BUGS & IMPROVEMENTS

- [X] Message when pattern removed https://github.com/owncloud/android/pull/2083#issuecomment-369553395
- [X] Message when switching the security method https://github.com/owncloud/android/pull/2083#issuecomment-369558295
- [X] Landscape mode https://github.com/owncloud/android/pull/2083#issuecomment-369559560 [WONT FIX]

sd1998 commented:
This PR does not consist of the changes made for Issue #2027

jesmrec commented:
Thanks for contributing @sd1998 !! i have check a bit this feature and i suggest a couple of improvements:

- The section "Details" of the settings view should be renamed to "Security"
- The word "Lock" is uppercase in Pattern and lowercase in Passcode
- Pattern Lock can only be set if the passcode if not set. But if both are enabled, the pattern lock can not be disabled. Which priorities did you set for both security methods? 

CC @davigonz

sd1998 commented:
@jesmrec I will start working on the changes you have suggested.

jesmrec commented:
ok @sd1998. It is important to set a policy of priorities how the different "locks" are going to be handled for the user (third point of my previous message). Thinking also that other locks can be included in the future (for example, the fingerprint). 

iOS handle it (touch id) in the following way: passcode is always mandatory, and if it is enabled, then other locks are visible and can be enabled if the user wants to.

Which is your idea about?

sd1998 commented:
@jesmrec Actually the way I have implemented the feature both the type of locks cannot be set at the same time.

jesmrec commented:
ok, so you have a bug ;) if you enable first passcode, then you can enable pattern as well.

So, the policy is that only one lock should be enabled at time. What do you thing @davigonz ? maybe passcode should be always enabled as shortcut?

davigonz commented:
> Actually the way I have implemented the feature both the type of locks cannot be set at the same time

> So, the policy is that only one lock should be enabled at time.

@jesmrec @sd1998 

Well, for this cases, I think we should follow the same way in which the Android system protects our phone. 

Android allows us to set a pattern to lock the device without setting a passcode but forces us to set a passcode when using fingerprint for instance, so **passcode and pattern would be different ways to unlock ownCloud app and cannot be enabled at the same time**.

sd1998 commented:
@davigonz @jesmrec <br/> [comment](https://github.com/owncloud/android/pull/2083#issuecomment-356350004 ) <br/> Yes I agree with you and I implemented the feature so that any one of the locks can be used and not both of them together but I have no idea why it is not working I will fix this and send a PR.

sd1998 commented:
@jesmrec I have made the changes.

sd1998 commented:
@jesmrec @davigonz I implemented the feature in such a way that if the passcode was set then the pattern could not be added but I forgot to account for the other case(i.e if the pattern is set then the user should not be able to set passcode) hence the bug ([comment](https://github.com/owncloud/android/pull/2083#issuecomment-356350004)). But now I have fixed it.

jesmrec commented:
Thanks @sd1998, we will pass code review and QA, and we will let you know if we find some issues.

davigonz commented:
Now that this branch is out-of-date, please use what I described in https://github.com/owncloud/android/pull/2075#issuecomment-359428847

jesmrec commented:
This is ready for QA, isn't ir @sd1998?

sd1998 commented:
@davigonz Whenever I rebase with master of ownCloud should I create a new PR?

davigonz commented:
> Whenever I rebase with master of ownCloud should I create a new PR?

@sd1998 I don't think so, if you are able to rebase it properly wouldn't be necessary and you have even created a new branch to implement the feature so 👍

sd1998 commented:
@davigonz I guess the rebase worked.

sd1998 commented:
@jesmrec Has the QA started for this and [PR](https://github.com/owncloud/android/pull/2097)?

jesmrec commented:
@sd1998 not yet, but i will take care next days. It will be included in 2.7.0 for sure.

davigonz commented:
> Has the QA started for this and PR? 
>  not yet, but i will take care next days. It will be included in 2.7.0 for sure.

@sd1998 @jesmrec It would be better to review a bit the code before starting QA

jesmrec commented:
of course @davigonz, i will be ready for your signal ;)

davigonz commented on res/values/strings.xml:
```diff
@@ -28,6 +28,7 @@
     <string name="drawer_open">Open</string>
     <string name="prefs_category_general">General</string>
     <string name="prefs_category_more">More</string>
+    <string name="prefs_category_security"> Security </string>
```
You do not need the extra white spaces at the beginning and end in these string resources.

davigonz commented on res/values/strings.xml:
```diff
@@ -46,6 +47,19 @@
+    <string name="pattern_enter_pattern"> Please enter your pattern </string>
+    <string name="pattern_configure_pattern"> Enter your pattern </string>
+    <string name="pattern_remove_pattern"> Remove your pattern </string>
+    <string name="pattern_incorrect_pattern"> Incorrect Pattern </string>
```
Pattern would better be lowercase

davigonz commented on res/values/strings.xml:
```diff
@@ -46,6 +47,19 @@
+    <string name="pattern_incorrect_pattern"> Incorrect Pattern </string>
+    <string name="pattern_reenter_pattern"> Please re-enter your pattern </string>
+    <string name="pattern_not_same_pattern"> The patterns are not same </string>
+    <string name="pattern_stored"> Pattern Stored </string>
```
Stored would better be lowercase

davigonz commented on res/values/strings.xml:
```diff
@@ -46,6 +47,19 @@
+    <string name="pattern_reenter_pattern"> Please re-enter your pattern </string>
+    <string name="pattern_not_same_pattern"> The patterns are not same </string>
+    <string name="pattern_stored"> Pattern Stored </string>
+    <string name="pattern_removed"> Pattern Removed </string>
```
Removed would better be lowercase

davigonz commented on res/values/strings.xml:
```diff
@@ -46,6 +47,19 @@
+    <string name="pattern_not_same_pattern"> The patterns are not same </string>
+    <string name="pattern_stored"> Pattern Stored </string>
+    <string name="pattern_removed"> Pattern Removed </string>
+    <string name="passcode_already_set"> Passcode Already Set </string>
```
Already and Set would better be lowercase

davigonz commented on res/values/strings.xml:
```diff
@@ -46,6 +47,19 @@
+    <string name="pattern_stored"> Pattern Stored </string>
+    <string name="pattern_removed"> Pattern Removed </string>
+    <string name="passcode_already_set"> Passcode Already Set </string>
+    <string name="pattern_already_set"> Pattern Already Set </string>
```
Same here

davigonz commented on build.gradle:
```diff
@@ -46,15 +46,16 @@ dependencies {
     compile "com.android.support:design:${supportLibraryVersion}"
     compile 'com.jakewharton:disklrucache:2.0.2'
     compile 'com.google.android.exoplayer:exoplayer:r2.2.0'
+    compile 'com.andrognito.patternlockview:patternlockview:1.0.0'
```
When we use a third-party library we usually include the information in our [THIRD_PARTY.txt](https://github.com/owncloud/android/blob/master/THIRD_PARTY.txt) file. Could you do the same with this library? You can have a look at the format we followed with other libraries in that file.

sd1998 replied:
As in this picture for the floating action button dependency it is written that it is placed in libs/com-getbase-floatingactionbutton-1-10-0-exploded-aar. Similarly what should I write for the placed at point (for the pattern lock view dependency).What I mean is what does placed at signify.</br>
<img width="1013" alt="screen shot 2018-02-05 at 9 23 36 pm" src="https://user-images.githubusercontent.com/22047421/35813943-e3c50840-0aba-11e8-8d7a-2cd542fe6ab8.png">

davigonz replied:
Forget this change for the moment, it seems that our THIRD_PARTY.txt is not properly updated, I will ping you when is ready.

sd1998 replied:
@davigonz Then I will commit the other changes. Thanks.

davigonz commented on src/com/owncloud/android/authentication/PatternManager.java:
```diff
@@ -0,0 +1,97 @@
+import java.util.Set;
+
+/**
+ * Created by shashvatkedia on 21/12/17.
```
Given that this is a new class, please delete this line and add a new comment in the top with the authors and copyright information:

```
/**
 *   ownCloud Android client application
 *
 *   @author YOUR NAME HERE
 *   Copyright (C) 2018 ownCloud GmbH.
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License version 2,
 *   as published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
```

davigonz commented:
@sd1998 I've just started with the code review, some changes are requested.

sd1998 commented on AndroidManifest.xml:
```diff
@@ -61,6 +61,8 @@
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
     <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
     <uses-permission android:name="android.permission.WAKE_LOCK" />
+    <uses-permission android:name="android.permission.READ_PHONE_STATE" />
```
@davigonz I added them when I working on the Issue #2027 (when I was working on master rather than creating a new branch) but I think I forgot to remove it during rebase. I will remove them now.

sd1998 commented on AndroidManifest.xml:
```diff
@@ -261,11 +258,14 @@
-        </activity>
+            android:theme="@style/Theme.ownCloud.Fullscreen" />
+        <activity android:name=".ui.activity.CapturedFileChecker" />
+        <activity android:name=".ui.activity.Camera" />
```
Yes I have removed them.

sd1998 commented:
@davigonz I have done all the changes other than [this](https://github.com/owncloud/android/pull/2083#discussion_r165982006).

sd1998 commented:
@davigonz  Please help me with [this](https://github.com/owncloud/android/pull/2083#discussion_r165982006).

davigonz commented:
> Please help me with this.

@sd1998 Resuming the code review here...

davigonz commented on src/com/owncloud/android/authentication/PatternManager.java:
```diff
@@ -0,0 +1,97 @@
+        visibleAcctivitiesCounter++;
+    }
+
+    private boolean patternShouldBeRequested(){
```
Please move this method and patternIsEnabled to the end of the class, similar to PassCodeManager class

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,270 @@
+package com.owncloud.android.ui.activity;
```
Include the same as https://github.com/owncloud/android/pull/2083/files#r165983321

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,270 @@
+        if (!BuildConfig.DEBUG) {
+            getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);
+        }
+            setContentView(R.layout.activity_pattern_lock);
```
Indent this line properly

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,270 @@
+    private static String COUNT_VALUE = "COUNT_VALUE";
+
+    private boolean patternPresent = false;
+    private boolean patternExpViewState = false;
```
This is only used in onCreate method, so you wouldn't need a global variable

sd1998 replied:
@davigonz Actually I am using these variables in many methods not just onCreate() hence they should be global.

davigonz replied:
I cannot see where are patternExpViewState and patternHeaderViewText variables being used, apart from onCreate() method

sd1998 replied:
@davigonz Yes I did not check for all the variables that I have declared. I will fix it now.

sd1998 replied:
@davigonz Yes I did not check for all the variables that I have declared. I will fix it now.

davigonz replied:
Thanks @sd1998 , whenever you can, no hurries

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,270 @@
+    private boolean patternExpViewState = false;
+    private String patternValue;
+    private String newPatternValue;
+    private String patternHeaderViewText;
```
Same here

sd1998 replied:
[this](https://github.com/owncloud/android/pull/2083#discussion_r167548081)

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,270 @@
+
+import android.content.Intent;
+import android.content.SharedPreferences;
+import android.content.res.Configuration;
```
Import not needed

sd1998 replied:
Yes I have removed them.

davigonz commented:
@sd1998 Thanks for applying the changes I requested the other day, I've just included a few more in the second code review

sd1998 commented on res/layout-land/activity_pattern_lock.xml:
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="utf-8"?>
```
@davigonz Yes in some devices the cancel button does not fit in landscape orientation. So should I remove this view?

davigonz replied:
Ok, if it does not fit in some devices, keep this landscape view 👍

sd1998 commented:
@davigonz I have made all the changes.

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,275 @@
+            cancelButton.setOnClickListener(null);
+        }
+    }
+
```
Too much space here

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,275 @@
+            @Override
+            public void onComplete(List<PatternLockView.Dot> list) {
+                if (ACTION_REQUEST_WITH_RESULT.equals(getIntent().getAction())) {
+                    if (count == 0) {
```
What is this count for? Maybe a comment here?

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,275 @@
+                patternPresent = savedInstanceState.getBoolean(KEY_CONFIRMING_PATTERN);
+                patternValue = savedInstanceState.getString(KEY_PATTERN_STRING);
+                patternHeaderViewText = savedInstanceState.getString(PATTERN_HEADER_VIEW_TEXT);
+                patternExpViewState = savedInstanceState.getBoolean(PATTERN_EXP_VIEW_STATE);
```
I do not understand exactly what is patternExpViewState about? A comment here?

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,275 @@
+        }
+        setContentView(R.layout.activity_pattern_lock);
+        patternHeaderView = (TextView) findViewById(R.id.header_pattern);
+        patternExpView = (TextView) findViewById(R.id.explanation_pattern);
```
Please use **m** character before the name of the variables, just to be consistent with the other classes we have in our code, e.g. mPatternExpView. Anyway, I would use mPatternExplanation and mPatternHeader, IMO, View suffix doesn't give us valuable information about the variable.

sd1998 replied:
I have change the variable names and added <b>m</b> as a prefix

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,275 @@
+        cancelButton = (Button) findViewById(R.id.cancel_pattern);
+        patternLockView = (PatternLockView) findViewById(R.id.pattern_lock_view);
+        patternLockView.clearPattern();
+        if (ACTION_CHECK.equals(getIntent().getAction())) {
```
I miss some comments here to explain the different actions in this activity. I guess this activity is being used to request the pattern the first time user enables the feature and to check if is correct when opening the app. Can you include a comment for each if else case explaining a bit that?

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    private void processPattern() {
+        if (ACTION_CHECK.equals(getIntent().getAction())) {
```
Some comments here also

davigonz commented on src/com/owncloud/android/ui/activity/Preferences.java:
```diff
@@ -139,21 +144,49 @@ public void onCreate(Bundle savedInstanceState) {
+                        intent.setAction(state ? PatternLockActivity.ACTION_REQUEST_WITH_RESULT :
+                                PatternLockActivity.ACTION_CHECK_WITH_RESULT);
+                        startActivityForResult(intent, state ? ACTION_REQUEST_PATTERN :
+                                ACTION_CONFIRM_PATTERN);
```
This fits in the previous line

davigonz commented on src/com/owncloud/android/ui/activity/Preferences.java:
```diff
@@ -77,14 +77,19 @@
+    private static final int ACTION_CONFIRM_PATTERN = 8;
 
     private CheckBoxPreference pCode;
+    private CheckBoxPreference patternCode;
```
This variable should better not include Code but something like pPattern. I think that **p** character means preference in this case, have a look at the variable just below `Preference pAboutApp`

davigonz commented on src/com/owncloud/android/ui/activity/Preferences.java:
```diff
@@ -514,6 +546,8 @@ protected void onResume() {
                 PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
         boolean state = appPrefs.getBoolean(PassCodeActivity.PREFERENCE_SET_PASSCODE, false);
```
Please rename this variable as passCodeState, to distinguish it better from patternState

davigonz commented on src/com/owncloud/android/ui/activity/Preferences.java:
```diff
@@ -609,6 +643,26 @@ protected void onActivityResult(int requestCode, int resultCode, Intent data) {
             }
         }
+        else if(requestCode == ACTION_REQUEST_PATTERN && resultCode == RESULT_OK){
+            String patterValue = data.getStringExtra(PatternLockActivity.KEY_PATTERN);
```
You have missed the n of pattern here

davigonz commented:
@sd1998 I finished the code review with some requested changes

davigonz commented:
Hi @sd1998 , I've seen you are interested in other new features, but it would be better to finish this before, do you have any blocker with the requested changes? If so, ask me to move forward this feature, thanks.

sd1998 commented:
@davigonz  I have made the changes.

sd1998 commented:
@davigonz are there any more changes required? And can you tell me more about the ownCloud bounty program? Thanks.

davigonz commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -0,0 +1,297 @@
+                     * the pattern value input the second time (which is stored in newPatternValue) if both the variables hold the same value
+                     * then the pattern is set.
+                     */
+                    if (mCount == 0) {
```
Do you really need this counter? You could check the second time whether mPatternValue is not empty and set the input to mNewPatternValue in that case.

sd1998 replied:
@davigonz Yes that can be done. I will do that. Thanks.

davigonz commented:
> are there any more changes required? And can you tell me more about the ownCloud bounty program? Thanks.

Hi @sd1998, I requested a few more minor changes. About the ownCloud bounty program, is pretty easy, we have the next repository https://www.bountysource.com/teams/owncloud/issues?tracker_ids=11489 where people put issues they would like to have implemented in ownCloud Android app and pay a bounty for it.

sd1998 commented:
@davigonz I will work on the changes.

sd1998 commented on src/com/owncloud/android/ui/activity/PatternLockActivity.java:
```diff
@@ -76,37 +73,42 @@ protected void onCreate(Bundle savedInstanceState) {
+        * patternExpViewState holds the boolean value that signifies weather the patternExpView should be visible or not.
+        * it is set to true when the pattern is set and when the pattern is removed.
+         */
+        boolean mPatternExpViewState = false;
```
Yes that would be a better option I have changed the variable name.

sd1998 replied:
Yes that will be a better option I have changed the variable name.

sd1998 commented:
@davigonz I have made the changes.

davigonz commented:
> I have made the changes.

Awesome @sd1998 , thank you very much, changes approved, let's start with QA @jesmrec

sd1998 commented:
@davigonz what about the THIRD_PARTY.txt file how should I update it?

davigonz commented:
> what about the THIRD_PARTY.txt file how should I update it?

Hi @sd1998 , have a look at one of the libraries I recently added [here](https://github.com/owncloud/android/blob/gif_support/THIRD_PARTY.txt#L67). 

You can include in THIRD_PARTY.txt file something like:

>     PatternLockView, version 1.0.0
>     Copyright (C) 2017 aritraroy
>     Licensed under Apache License, Version 2.0.
>     Included in the ownCloud client APK.
>     See https://github.com/aritraroy/PatternLockView

Another important thing, please point the owncloud-android-library to commit `0e7fd8ee8fb1ad082acba8911412571b206ee1f0`, this branch is pointing to commit `7c7cad7e990705c180f4fe2934191c317a3f72fc`. 

When you navigate between different branches, I recommend you to use `git submodule update` to update the library version that is being pointed from the current app branch.

Just one last thing before starting QA, please rebase the branch with master.

Thanks!

sd1998 commented:
@davigonz I have updated the THIRD_PARTY.txt and to point the owncloud-android-library to 0e7fd8ee8fb1ad082acba8911412571b206ee1f0 I used #2075 [comment](https://github.com/owncloud/android/pull/2097#issuecomment-360198056).

jesmrec commented:
@sd1998 let's go for it!!

jesmrec commented:
### IMPROVEMENT 

Steps:
1. Add pattern lock
2. Click on pattern lock in order to remove it

Current: Message `The pattern will be requested every time the app is started`
Expected: `The pattern will not be asked anymore`or similar. The current message fits when you set a new pattern lock, but not much if you remove it.

Tested with Nexus 6P v7

jesmrec commented:
### IMPROVEMENT 

Steps:
1. Add pattern lock
2. Try to enable passcode (or viceversa)

Current: Message `Pattern already set` (or `Passcode already set`) in a Snackbar

This message does not clarify the fact that only one of these methods can be enabled at time.

Suggestions:

- `Please, disable XXX first`  (where XXX is the enabled method)
- ...

(any other @davigonz ? )

Tested with Nexus 6P v7

jesmrec commented:
### IMPROVEMENT 

Not sure if this one depends on the library you used to develop the feature

Steps:
1. Add pattern lock
2. Set the device in landscape in the pattern lock view

Current: Cancel button is shown at the right of the pattern points
Expected: Cancel button is shown below, as in landscape.

Tested with Nexus 6P v7

sd1998 commented:
@jesmrec [this](https://github.com/mineshmathew/OpenDocScanner) does not depend on the library I have used. I placed the cancel button to the right of the patternLockView because in landscape orientation the cancel button does not fit properly bellow the patternLockView (i.e due to the margin between the button and the epatternLockView the cancel button is not properly visible on my device (Oneplus 3T) hence I placed it beside the patternLockView).

davigonz commented:
> Please, disable XXX first (where XXX is the enabled method)
> (any other @davigonz ? )

@sd1998 

Yes, or something like: "Pattern and passcode locks cannot be enabled at the same time, please disable XXX first"

jesmrec commented:
QA is passed with the suggested improvements. Functionally is working pretty fine (congrats for the work!!), so we can discuss the improvements, and then, after checking them and rebasing the branch, it will be ready to merge.

sd1998 commented:
@jesmrec @davigonz I have made the changes and done the rebase.

jesmrec commented:
Changes tested and all OK.

Approved

davigonz commented:
> Changes tested and all OK.
> Approved

Awesome, merging it... Thanks @sd1998 for this feature
